---
title: "DMR_stats_annot"
author: "Javier Rodriguez Casariego"
date: "03/08/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/JARC_2T/Data_PhD/Rapid_Project/Clonal_divergence/ACER_clonal_divergence/WGBS')
```

load libraries
```{r}
library(data.table)
library(gplots)
library(ggplot2) #install.packages('ggplot2')
library(dplyr)
library(broom)
library(RColorBrewer)
library(egg) #install.packages('egg')
library(purrr)
library(nlme)
library(vegan)
```

```{r set functions}
## ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10) %+replace%    #, base_family = "Arial"
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA),
      axis.text.x = element_text(angle=45, hjust=1, vjust = 1)#,
      #legend.title = element_text(size = 8), 
      #legend.text = element_text(size = 7)
    )
}

```
# Part 1:  Filter methylated regions for regions that have coverage in 75% of samples per group

read in data
```{r load data}
meta <- read.csv("data/Sample_Metadata.csv", header = TRUE)
colnames(meta)[1] <- "site"

all_DMR_100bp<-read.csv("data/DMR_files/AllDepths_DMR100bp_cov10x_rms_results_collapsed.tsv", sep = "\t", header = TRUE)
colnames(all_DMR_100bp)[1] <- "chr"

all_DMR_250bp<-read.csv("data/DMR_files/AllDepths_DMR250bp_cov10x_rms_results_collapsed.tsv", sep = "\t", header = TRUE)
colnames(all_DMR_250bp)[1] <- "chr"

```

```{r prepare data for analysis}
###Filter regions present in over 75% of samples

df <- data.frame() #create empty data frame to bind filtered rows into
for(i in (1:nrow(all_DMR_100bp))){
  shallow <- all_DMR_100bp[i,c(7,9,11,13,15,17,19,21,23,25,27,29,31)] #define columns from the category shallow
  deep <- all_DMR_100bp[i,c(8,10,12,14,16,18,20,22,24,26,28,30,32)] #define columns from the category deep
      if(length(which(is.na(shallow))) < 2 & length(which(is.na(deep))) < 2){
  df <- rbind(df,all_DMR_100bp[i,]) #conditional statement: if less than 2 samples/category have NA for % methylation bind the whole row to the new dataframe
  }
}
#save df as a specific variable
all_DMR_100bp <- df

all_DMR_100bp$ID <- paste(all_DMR_100bp$chr,":", all_DMR_100bp$start,"-",all_DMR_100bp$end, sep = "")
all_DMR_100bp$ID <- gsub("__.*__.*:",":",all_DMR_100bp$ID)

df <- data.frame() #create empty data frame to bind filtered rows into
for(i in (1:nrow(all_DMR_250bp))){
  shallow <- all_DMR_250bp[i,c(7,9,11,13,15,17,19,21,23,25,27,29,31)] #define columns from the category shallow
  deep <- all_DMR_250bp[i,c(8,10,12,14,16,18,20,22,24,26,28,30,32)] #define columns from the category deep
      if(length(which(is.na(shallow))) < 2 & length(which(is.na(deep))) < 2){
  df <- rbind(df,all_DMR_250bp[i,]) #conditional statement: if less than 2 samples/category have NA for % methylation bind the whole row to the new dataframe
  }
}
#save df as a specific variable
all_DMR_250bp <- df

all_DMR_250bp$ID <- paste(all_DMR_250bp$chr,":", all_DMR_250bp$start,"-",all_DMR_250bp$end, sep = "")
all_DMR_250bp$ID <- gsub("__.*__.*:",":",all_DMR_250bp$ID)


#build %meth per DMR matrix for 100bp dataset
perc_methDMR_100 <- all_DMR_100bp[,c(33,7:32)]
colnames(perc_methDMR_100) <- gsub("methylation_level_","", colnames(perc_methDMR_100))

perc_meth_100 <- perc_methDMR_100[,-1] %>%
    dplyr::select(sort(names(.)))
boxplot(perc_meth_100) + theme_custom()
row.names(perc_meth_100) <- perc_methDMR_100$ID 

#build %meth per DMR matrix for 250bp dataset
perc_methDMR_250 <- all_DMR_250bp[,c(33,7:32)]
colnames(perc_methDMR_250) <- gsub("methylation_level_","", colnames(perc_methDMR_250))

perc_meth_250 <- perc_methDMR_250[,-1] %>%
    dplyr::select(sort(names(.)))

row.names(perc_meth_250) <- perc_methDMR_250$ID 

# Build dataset for general stats 

perc_meth_by_sample_100 <- tidyr::gather(all_DMR_100bp[,7:ncol(all_DMR_100bp)], "Sample.ID", "perc.meth",1:26) %>%
  tidyr::spread(ID, perc.meth)
perc_meth_by_sample_100$Sample.ID <- gsub("methylation_level_","", perc_meth_by_sample_100$Sample.ID)

perc_meth_by_sample100 <- merge(meta[,c(1,4:5)], perc_meth_by_sample_100, by = "Sample.ID")

```
## Run permanova to evaluate methylation interaction with site and genotype

```{r}
set.seed(1234)
adonis_site <- adonis (perc_meth_by_sample100[, c(4:ncol(perc_meth_by_sample100))] ~ site, strata = perc_meth_by_sample100$genotype, data = perc_meth_by_sample100, permutations = 9999, method = "binomial", na.rm = TRUE)

adonis_siteGenot <- adonis (perc_meth_by_sample100[, c(4:ncol(perc_meth_by_sample100))] ~ site*genotype, data = perc_meth_by_sample100, permutations = 9999, method = "manhattan",  na.rm = TRUE)

adonis_site
adonis_siteGenot

```


# Part 2: Run group statistics on regions to find regions that are significantly different among groups

reformat data for calculating group effect
```{r}
#reformat all for t test
all_DMR_100_STACKED <- tidyr::gather(all_DMR_100bp[,7:ncol(all_DMR_100bp)], "Sample.ID", "perc.meth",1:26)

#simplify sample name
all_DMR_100_STACKED$Sample.ID <- gsub("methylation_level_","", all_DMR_100_STACKED$Sample.ID)

#merge with meta data
all_DMR_100_STACKED <- merge(all_DMR_100_STACKED, meta, by = "Sample.ID")

```

plot % meth dist.
```{r}
a <- ggplot(all_DMR_100_STACKED) + 
  geom_histogram(aes(perc.meth, group = site, color = site,fill = site), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

a
```

arc sin sqrt transform the data
```{r}
#arcsin sqrt transformation function
asinTransform <- function(p) { asin(sqrt(p))}

all_DMR_100_STACKED_asin <- all_DMR_100_STACKED
all_DMR_100_STACKED_asin$perc.meth <- asinTransform(all_DMR_100_STACKED_asin$perc.meth)
```

plot transformed dist.
```{r}
b <- ggplot(all_DMR_100_STACKED_asin) + 
  geom_histogram(aes(perc.meth, group = site, color = site,fill = site), bins = 20, position = "identity", alpha = 0.5) + 
  theme_bw() + 
  xlab("transformed fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

b

jpeg("output/DMR_100_percMeth_histogram.jpeg", width = 8, height = 4, units = "in", res = 300)
ggarrange(a, b, nrow = 1)
dev.off()

```
## Run anova on TRANSFORMED data to assess group differences for each DMR

## One-way anova
**Hypotehsis: there is no effect of outplanting site (here site)
```{r}
all_DMR_100_1way_aov <- all_DMR_100_STACKED_asin %>% group_by(ID) %>%
do(meth_aov_models = aov(perc.meth~site, data =  . ))
#summarize ANOVA data
all_DMR_100_1way_aov_modelsumm  <- all_DMR_100_1way_aov %>% ungroup %>% 
    pull(meth_aov_models) %>% 
    map_dfr(tidy, .id = 'grp')
#spread out pvalues
all_DMR_100_1way_aov_modelsumm_wide <- data.frame(tidyr::pivot_wider(all_DMR_100_1way_aov_modelsumm, names_from = term, values_from = c("df", "sumsq", "meansq","statistic","p.value"),names_sep = "_" ))
all_DMR_100_1way_aov_modelsumm_wide <- cbind(all_DMR_100_1way_aov[,1], all_DMR_100_1way_aov_modelsumm_wide[,-c(1)])
length(all_DMR_100_1way_aov_modelsumm_wide[which(all_DMR_100_1way_aov_modelsumm_wide$p.value_site <= 0.05), 1])

```

## Two-way anova
**Hypotehsis: there is no effect of outplanting site (here site), colony or its interaction
```{r}
all_DMR_100_2way_aov <- all_DMR_100_STACKED_asin %>% group_by(ID) %>%
do(meth_aov_models = aov(perc.meth ~site*genotype, data =  . ))
#summarize ANOVA data
all_DMR_100_2way_aov_modelsumm  <- all_DMR_100_2way_aov %>% ungroup %>% 
    pull(meth_aov_models) %>% 
    map_dfr(tidy, .id = 'grp')

#spread out pvalues
all_DMR_100_2way_aov_modelsumm_wide <- data.frame(tidyr::pivot_wider(all_DMR_100_2way_aov_modelsumm, names_from = term, values_from = c("df", "sumsq", "meansq","statistic","p.value"),names_sep = "_" ))
all_DMR_100_2way_aov_modelsumm_wide <- cbind(all_DMR_100_2way_aov[,1], all_DMR_100_2way_aov_modelsumm_wide[,-c(1)])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_site <= 0.05), 1])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_genotype <= 0.05), 1])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_site.genotype <= 0.05), 1])
```

plot heatmap of Anova for site p.val > 0.05 sig DMRs
```{r}
#create matrix for one way ANOVA results
aov_0.05_site_DMR_m <- as.matrix(all_DMR_100bp[,7:32])
rownames(aov_0.05_site_DMR_m) <- all_DMR_100bp$ID
aov_0.05_site_DMR_m <- aov_0.05_site_DMR_m[which(rownames(aov_0.05_site_DMR_m) %in% pull(all_DMR_100_1way_aov_modelsumm_wide[which(all_DMR_100_1way_aov_modelsumm_wide$p.value_site<= 0.05),],ID)),]

#create matrix for two way ANOVA results
aov2_0.05_site_DMR_m <- as.matrix(all_DMR_100bp[,7:32])
rownames(aov2_0.05_site_DMR_m) <- all_DMR_100bp$ID
aov2_0.05_site_DMR_m <- aov2_0.05_site_DMR_m[which(rownames(aov2_0.05_site_DMR_m) %in% pull(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_site.genotype<= 0.05),],ID)),]


#remove extra text from column names
colnames(aov_0.05_site_DMR_m) <- gsub("methylation_level_","",colnames(aov_0.05_site_DMR_m))
colnames(aov2_0.05_site_DMR_m) <- gsub("methylation_level_","",colnames(aov2_0.05_site_DMR_m))

###Visualize group means fo all
#calculate group means
MeanDeep <- rowMeans(aov_0.05_site_DMR_m[,grep("_D", colnames(aov_0.05_site_DMR_m))], na.rm = TRUE)
MeanShallow <- rowMeans(aov_0.05_site_DMR_m[,grep("_S", colnames(aov_0.05_site_DMR_m))], na.rm = TRUE)

#bind all group means together
aov_0.05_site_DMR_mean <- as.matrix(data.frame(cbind(MeanDeep,MeanShallow)))

#plot site mean comparison
colnames(aov_0.05_site_DMR_mean) <- c("Deep", "Shallow")
jpeg("output/all.aov_0.05_100_DMR_heatmap.jpg", width = 600, height = 1000)
heatmap.2(aov_0.05_site_DMR_mean,margins = c(10,20), cexCol = 2, distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=NA, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(aov_0.05_site_DMR_mean),rowsep=1:nrow(aov_0.05_site_DMR_mean),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()

#plot each individual sample

genet_color <- data.frame(sample=colnames(aov_0.05_site_DMR_m), genet=meta[meta$Sample.ID %in% genet_color$sample, 5], color = color <-c("#F8766D", "#F8766D", "#FF64B0", "#FF64B0", "#00B4F0", "#00B4F0", "#00C08B", "#00C08B", "#619CFF", "#619CFF", "#F564E3", "#F564E3", "#DE8C00", "#DE8C00", "#C77CFF", "#C77CFF", "#00BA38", "#00BA38", "#DE8C00", "#DE8C00", "#00BFC4", "#00BFC4", "#B79F00", "#B79F00", "#7CAE00", "#7CAE00"))


# set the custom distance and clustering functions
hclustfunc <- function(x) hclust(x,method = 'average')
distfunc <- function(x) as.dist(1 - cor(t(x), use = "pa"))
key.xtickfun <- function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)
  }

jpeg("DMR.aov_0.05_100_bysample_heatmap.jpg", width = 1000, height = 1000)
heatmap.2(aov_0.05_site_DMR_m, margins = c(10,10), hclustfun=hclustfunc, distfun=distfunc, key.xtickfun = key.xtickfun, cexCol = 2, Colv=T, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(aov_0.05_site_DMR_m),rowsep=1:nrow(aov_0.05_site_DMR_m),keysize=0.5, key.par = list(cex=1), ColSideColors=genet_color$color)
dev.off()


```
#write out DMR bed file for 1way anova

```{r}
all_DMR_100_1way_aov_modelsumm_wide$scaffold <- gsub("\\:.*","", all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide$start <- gsub(".*\\:","", all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide$start <- gsub("-.*","", all_DMR_100_1way_aov_modelsumm_wide$start)
all_DMR_100_1way_aov_modelsumm_wide$end <- gsub(".*-","",  all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide <- all_DMR_100_1way_aov_modelsumm_wide[,c(1,(ncol(all_DMR_100_1way_aov_modelsumm_wide)-2):ncol(all_DMR_100_1way_aov_modelsumm_wide),2:(ncol(all_DMR_100_1way_aov_modelsumm_wide)-3))]
#order by pvalue
all_DMR_100_1way_aov_modelsumm_wide <- all_DMR_100_1way_aov_modelsumm_wide[order(all_DMR_100_1way_aov_modelsumm_wide$p.value_site),]
write.csv(all_DMR_100_1way_aov_modelsumm_wide,"output/aov_site.csv", row.names = FALSE, quote = FALSE)
```

### Export file for feature analysis

generate bedfiles to be used in bedtools intersect to determine features that overlap with regions
```{r}
#subset model data for scaffold start and end positions for DMRs significant at p < 0.05
DMR_100_sig.bed <- all_DMR_100_1way_aov_modelsumm_wide[which(all_DMR_100_1way_aov_modelsumm_wide$p.value_site < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig.bed$test <- "site"

#write out data
write.table(DMR_100_sig.bed, "output/aov_0.05_site.bed", sep = "\t",row.names = F, col.names = F, quote = F)

```

#write out DMR bed file for 2way anova

```{r}
all_DMR_100_2way_aov_modelsumm_wide$scaffold <- gsub("\\:.*","", all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide$start <- gsub(".*\\:","", all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide$start <- gsub("-.*","", all_DMR_100_2way_aov_modelsumm_wide$start)
all_DMR_100_2way_aov_modelsumm_wide$end <- gsub(".*-","",  all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide <- all_DMR_100_2way_aov_modelsumm_wide[,c(1,(ncol(all_DMR_100_2way_aov_modelsumm_wide)-2):ncol(all_DMR_100_2way_aov_modelsumm_wide),2:(ncol(all_DMR_100_2way_aov_modelsumm_wide)-3))]
#order by pvalue
all_DMR_100_2way_aov_modelsumm_wide <- all_DMR_100_2way_aov_modelsumm_wide[order(all_DMR_100_2way_aov_modelsumm_wide$p.value_site),]
write.csv(all_DMR_100_2way_aov_modelsumm_wide,"output/aov2way_all.csv", row.names = FALSE, quote = FALSE)
```

### Export file for feature analysis

generate bedfiles to be used in bedtools intersect to determine features that overlap with regions
```{r}
#subset model data for scaffold start and end positions for DMRs significant at p < 0.05
DMR_100_sig_site_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_site < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_site_2way.bed$test <- "site"

#write out data
write.table(DMR_100_sig_site_2way.bed, "output/aov2way_0.05_site.bed", sep = "\t",row.names = F, col.names = F, quote = F)

DMR_100_sig_genet_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_genotype < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_genet_2way.bed$test <- "genotype"

#write out data
write.table(DMR_100_sig_genet_2way.bed, "output/aov2way_0.05_genotype.bed", sep = "\t",row.names = F, col.names = F, quote = F)

DMR_100_sig_interaction_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_site.genotype < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_interaction_2way.bed$test <- "site:genet"

#write out data
write.table(DMR_100_sig_interaction_2way.bed, "output/aov2way_0.05_site:genet.bed", sep = "\t",row.names = F, col.names = F, quote = F)
```

##plot heatmap of sig DMRs AOV_1way overlaping with gene regions
```{r}
#create matrix for one way ANOVA results
gene_DMR_m <- as.matrix(all_DMR_100bp[,7:32])
rownames(gene_DMR_m) <- all_DMR_100bp$ID

gene_overlap <- read.csv("analyses/Genomic_features/DMR_genomic_features/aov_0.05_site_gene_trans.txt", sep = "\t", header = FALSE)
gene_overlap$ID <- paste(gene_overlap$V1,":", gene_overlap$V2,"-",gene_overlap$V3, sep = "")
gene_overlap$ID <- gsub("__.*__.*:",":",gene_overlap$ID)

gene_DMR_m <- gene_DMR_m[which(rownames(gene_DMR_m) %in% gene_overlap$ID),]

#remove extra text from column names
colnames(gene_DMR_m) <- gsub("methylation_level_","",colnames(gene_DMR_m))

###Visualize group means fo all
#calculate group means
MeanDeep <- rowMeans(gene_DMR_m[,grep("_D", colnames(gene_DMR_m))], na.rm = TRUE)
MeanShallow <- rowMeans(gene_DMR_m[,grep("_S", colnames(gene_DMR_m))], na.rm = TRUE)

#bind all group means together
gene_DMR_mean <- as.matrix(data.frame(cbind(MeanDeep,MeanShallow)))

#plot site mean comparison
colnames(gene_DMR_mean) <- c("Deep", "Shallow")
jpeg("output/gene.aov_0.05_100_DMR_heatmap.jpg", width = 600, height = 1000)
heatmap.2(gene_DMR_mean,margins = c(10,20), cexCol = 2, distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=NA, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(gene_DMR_mean),rowsep=1:nrow(gene_DMR_mean),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()

#plot each individual sample
jpeg("output/DMR.gene_aov_0.05_100_bysample_heatmap.jpg", width = 1000, height = 1000)
heatmap.2(gene_DMR_m,margins = c(10,10), cexCol = 2, distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=TRUE, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(gene_DMR_m),rowsep=1:nrow(gene_DMR_m),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()

```